roles/openshift_hosted/defaults/main.yml

```diff
@@ -46,4 +46,4 @@ r_openshift_hosted_registry_os_firewall_deny: []
 r_openshift_hosted_registry_os_firewall_allow:
 - service: Docker Registry Port
   port: 5000/tcp
-  cond: "{{ r_openshift_hosted_use_calico }}"
+  cond: "{{ openshift_hosted_use_calico }}"
```

Comment:
I think this is what @sdodson was referencing when he said "It goes against some past architecture decisions such as requiring that all role variables are passed in but it's not like this is introducing more of that it's just substituting openshift_facts magic with more defaults."

I do prefer that we don't reference global variables like this.

Instead of just being able to call this role and pass in variables, I now need to make sure `openshift_hosted_use_calico` is set somewhere before I call this role.

Also, to discover this, I need to read through the code of the role.

Still feels like a bad pattern to me.

Reply:
Yeah, that's what I meant. Thanks for clarifying.

Reply:
openshift_hosted_use_calico is not a 'role variable,' as it's not specific to any one role.  It's a user-provided inventory variable that controls many roles.

What is the benefit of defining r_openshift_hosted_use_calico without a default in this role?  In reality, this variable is a direct derivative of openshift_hosted_use_calico.  r_openshift_hosted_use_calico provides no meaning without openshift_hosted_use_calico.

I think the r_variable pattern might have been adopted for a valid purpose, but I think it was too widely adopted and misapplied in cases like r_openshift_hosted_use_calico

Reply:
Good catch, I will need to provide default for openshift_hosted_use_calico here though.

Reply:
@mwoodson @twiest, would it be sufficient to place it in the defaults and have a sane default for the value?

Reply:
@michaelgugino "It's a user-provided inventory variable that controls many roles."

Yeah, that's what I was saying. That's why we call varialbes like that "global variables". Once you tie roles to that global variable, the role becomes much less flexible.

We prefer roles to be self contained. This includes not referencing global variables like this.

That way we can call this role and pass in what we'd like the role to use instead of having to ensure that a bunch of global variables are all set up.

Also, I'd prefer that the role variable be named for what is being enabled / disabled. The role is doing some logic with that variable, and thus we should name it based on what that logic actually means.

Then if we want the global variable to be used for that functionality, then we can pass it in for the role variable.

@sdodson Do you have the recording of the meeting where we discussed not using global variables in roles?

I think that would give @michaelgugino more context into the discussion / decision.

@kwoodson assuming you're talking about a role variable, then that sounds fine to me.

Reply:
@twiest This is not a 'global variable'.  It is an inventory variable.  Inventory variables override role defaults.  Each role that consumes such an inventory should also specify a role default, except in a very few special cases (e.g. passwords).

Imagine you're building a web app.  You need two roles, app, and mysql.  Does it make sense to give each role it's own unique name for mysql_port?  No, it doesn't.  We probably want mysql_port to be the same in each place.

Similarly, if we have a choice of different database backends, which affect both app and db role, the variable use_mysqldb would make sense.

If we were dynamically deriving some data and using set_fact to populate that data into a variable, then the method you are suggesting makes sense.  In the case we're currently discussing, it is adding zero-value and only introducing a middle-man variable with no value added, only negatives added.

> We prefer roles to be self contained. This includes not referencing global variables like this.
You're only trying to cover up the fact that an inventory variable is indeed being used.  The role is still 100% dependent on that inventory variable, whether you reference it directly or not.

Reply:
@michaelgugino 

> You're only trying to cover up the fact that an inventory variable is indeed being used. The role is still 100% dependent on that inventory variable, whether you reference it directly or not.

Well, the point is to decoupling the role from the inventory variable. This allows the role caller to specifically set the variable based on what they want done. The user of the role could use an inventory variable, or they could just set the role variable manually.

The point is, we're not forcing the caller of the role to have their inventory setup in a specific way before being able to call the role.

Also, again, in this specific case, that variable is determining if the registry should have a hole poked in iptables on port 5000. So it would be better to have a role variable with a name that specifically describes what it's being used for.  Something like `r_openshift_hosted_enable_registry_port_in_fw` or whatever (yes I changed the name of the var since my last comment).  :)

Reply:
@sdodson At this point, I think we should discuss this this in the arch meeting and get the rest of the team's opinions on this.

Would you mind adding it to the agenda?

Reply:
done.